src/main/java/net/earthcomputer/clientcommands/command/CHelpCommand.java

```diff
@@ -25,39 +29,92 @@ public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
+
+                List<String> commandNames = new ArrayList<String>();
+                for(CommandNode<ServerCommandSource> command : dispatcher.getRoot().getChildren())
+                {
+                    String commandName = command.getName();
+                    if(isClientSideCommand(commandName))
+                    {
+                        commandNames.add('/' + commandName);
+                        cmdCount++;
                     }
                 }
+
+                Paginator<String> paginator = new Paginator<String>(commandNames, 9);
+                Page<String> page = paginator.getPage(1);
+                for (int i = 0; i < page.Items.size(); i++)
+                {
+                    sendFeedback(page.Items.get(i));
+                }
+                sendFeedback(new TranslatableText("paginator.display.page", page.PageNumber, paginator.getPageCount()));
+
                 return cmdCount;
+
             })
-            .then(argument("command", greedyString())
+            .then(argument("page | command", greedyString())
                 .executes(ctx -> {
-                    String cmdName = getString(ctx, "command");
-                    if (!isClientSideCommand(cmdName))
-                        throw FAILED_EXCEPTION.create();
 
-                    ParseResults<ServerCommandSource> parseResults = dispatcher.parse(cmdName, ctx.getSource());
-                    if (parseResults.getContext().getNodes().isEmpty())
-                        throw FAILED_EXCEPTION.create();
+                    String userInput = getString(ctx, "page | command");
+
+                    if(isInteger(userInput))
+                    {
+                        int userPage = Integer.parseInt(userInput);
+
+                        List<String> commandNames = new ArrayList<String>();
+                        for(CommandNode<ServerCommandSource> command : dispatcher.getRoot().getChildren())
+                        {
+                            String commandName = command.getName();
+                            if(isClientSideCommand(commandName))
+                            {
+                                commandNames.add('/' + commandName);
+                            }
+                        }
+                        Paginator<String> paginator = new Paginator<String>(commandNames, 9);
+                        if(!paginator.isValidPage(userPage))
+                        {
+                            sendFeedback(new TranslatableText("paginator.incorrect.page"));
+                            return 0;
+                        }
+                        Page<String> page = paginator.getPage(userPage);
+                        for (int i = 0; i < page.Items.size(); i++)
+                        {
+                            sendFeedback(page.Items.get(i));
+                        }
+
+                        sendFeedback(new TranslatableText("paginator.display.page", page.PageNumber, paginator.getPageCount()));
```

Comment:
I think it would be better to include clickable navigation text on the page display.

Reply:
Ok.